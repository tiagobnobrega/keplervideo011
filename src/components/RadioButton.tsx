import { isEqual } from 'lodash';
import React from 'react';
import { StyleSheet, Text, View } from 'react-native';
import { COLORS } from '../styles/Colors';
import { scaleUxToDp } from '../utils/pixelUtils';
import FocusableElement from './FocusableElement';

const RadioButton = ({
  label,
  onSelect,
  selected,
  testID,
}: {
  label: string;
  onSelect: () => void;
  selected?: boolean;
  testID: string;
}) => {
  return (
    <View style={styles.buttonContainer} testID={testID}>
      <FocusableElement
        onPress={onSelect}
        onFocusOverrideStyle={styles.focusedView}
        style={[styles.radioButton, selected && styles.radioButtonSelected]}
        testID={`radio-btn-${label}`}
      />
      <Text style={styles.label} testID={`radio-text-${label}`}>
        {label}
      </Text>
    </View>
  );
};

const radioBtnUnchanged = (
  prevProps: {
    label: string;
    onSelect: () => void;
    selected?: boolean;
  },
  nextProps: {
    label: string;
    onSelect: () => void;
    selected?: boolean;
  },
) => {
  return isEqual(prevProps, nextProps);
};

export default React.memo(RadioButton, radioBtnUnchanged);

const styles = StyleSheet.create({
  buttonContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    paddingHorizontal: scaleUxToDp(15),
  },
  label: {
    fontSize: scaleUxToDp(27),
    color: COLORS.WHITE,
    marginLeft: scaleUxToDp(10),
  },
  focusedView: {
    borderColor: COLORS.YELLOW,
  },
  radioButton: {
    width: scaleUxToDp(30),
    height: scaleUxToDp(30),
    borderRadius: scaleUxToDp(15),
    borderWidth: 4,
    borderColor: COLORS.LIGHT_GRAY,
  },
  radioButtonSelected: {
    backgroundColor: COLORS.LIGHT_GRAY,
  },
});
