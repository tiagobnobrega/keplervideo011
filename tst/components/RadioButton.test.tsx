import { render } from '@testing-library/react-native';
import React from 'react';
import 'react-native';
import RadioButton from '../../src/components/RadioButton';

describe('RadioButton renders correctly', () => {
  it('renders correctly without styles', () => {
    const tree = render(
      <RadioButton label={''} onSelect={jest.fn()} testID={'radio-button-1'} />,
    );
    expect(tree).toMatchSnapshot();
  });
});

describe('RadioButton with React.memo', () => {
  it('renders only when props change', async () => {
    const label = 'test';
    const mockOnSelect = jest.fn();
    const { rerender } = render(
      <RadioButton
        label={''}
        onSelect={mockOnSelect}
        testID={'radio-button-1'}
      />,
    );
    const isEqualSpy = jest.spyOn(require('lodash'), 'isEqual');
    rerender(
      <RadioButton
        label={''}
        onSelect={mockOnSelect}
        testID={'radio-button-1'}
      />,
    );
    expect(isEqualSpy).toHaveBeenCalledTimes(1);
    rerender(
      <RadioButton
        label={label}
        onSelect={mockOnSelect}
        testID={'radio-button-1'}
      />,
    );
    expect(isEqualSpy).toHaveBeenCalledTimes(2);
    isEqualSpy.mockRestore();
  });
});
