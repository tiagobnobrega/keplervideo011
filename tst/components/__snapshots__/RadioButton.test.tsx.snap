// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButton renders correctly renders correctly without styles 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "paddingHorizontal": 15,
    }
  }
  testID="radio-button-1"
>
  <View
    accessibilityState={
      Object {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      Object {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "borderColor": "#D3D3D3",
        "borderRadius": 15,
        "borderWidth": 4,
        "height": 30,
        "opacity": 1,
        "width": 30,
      }
    }
    testID="radio-btn-"
  />
  <Text
    style={
      Object {
        "color": "#FFFFFF",
        "fontSize": 27,
        "marginLeft": 10,
      }
    }
    testID="radio-text-"
  />
</View>
`;
